text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,608 @@
+[RFC](https://github.com/DemiMarie/rfcs/blob/become/0000-proper-tail-calls.md) can be found at
+[DemiMarie/rust/tree/explicit-tailcalls](https://github.com/DemiMarie/rust/tree/explicit-tailcalls).
+A new implementation is planned as part of this RFC.
+
+The parser parses `become` exactly how it parses the `return` keyword. The difference in semantics is handled later.
+
+During type checking, the following are checked:
+
+1. The target of the tail call is, in fact, a simple call.
+2. The target of the tail call has the proper ABI.
+
+Later phases in the compiler assert that these requirements are met.
+
+New nodes are added in HIR and THIR to correspond to `become`. In MIR, the function call is checked that:
+1. The returned value is directly returned.
+2. There are no cleanups.
+3. The basic block being branched into has length zero.
+4. The basic block being branched into terminates with a return.
+
+If these conditions are fulfilled the function call following `become` is flagged to indicate the TCO requirement. This
+flag is then propagated to the corresponding backend. In the backend, there is an additional check if TCO can be
+performed.
+
+Should any check during compilation not pass a compiler error should be issued.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+<!-- Why should we *not* do this? -->
+As this feature should be mostly independent of other features the main drawback lies in the implementation and
+maintenance effort. This feature adds a new keyword which will need to be implemented not only in Rust but also in other
+tooling. The primary effort, however, lies in supporting this feature in the backends:
+- LLVM supports a `musttail` marker to indicate that TCO should be performed [docs](https://llvm.org/docs/LangRef.html#id327). Clang which already depends on this feature, seems to only generate correct code for the x86 backend [source](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1490009983) (as of 30.03.23).
+- GCC does not support an equivalent `musttail` marker.
+- WebAssembly accepted tail-calls into the [standard](https://github.com/WebAssembly/proposals/pull/157/) and Cranelift is now [working](https://github.com/bytecodealliance/rfcs/pull/29) towards supporting it.
+
+Additionally, this proposal is limited to exactly matching function signatures which will *not* allow general tail-calls, however, the work towards this initial version is likely to be useful for a more comprehensive version.
+
+There is also an unwanted interaction between TCO and debugging. As TCO by design elides stack frames this information is lost during debugging, that is the parent functions and their local variable values are incomplete. As TCO provides a semantic guarantee of constant stack usage it is also not generally possible to disable TCO for debugging builds as then the stack could overflow. (Still maybe a compiler flag could be provided to temporarily disable TCO for debugging builds.)
```

Comment:
Technically a debugger with proper Rust support could provide special support for tail calls in that it for example stores the callers for a chain of tail calls of up to N depth. Usually this should be enough to help with debugging since most often a bug is located close to the point of failure. Therefore you often don't need unlimited stack depth. Of course generally this is still a drawback but I think a much smaller one than pictured with proper debugger support.

Reply:
Good idea, I'll add your suggestion.